Урок 6. Задача look-alike

Домашнее задание

1. взять любой набор данных для бинарной классификации (можно скачать один с https://archive.ics.uci.edu/ml/datasets.php)

2. сделать feature engineering

3. обучить любой классификатор (какой вам нравится)

4. далее разделить ваш набор данных на два множества: P (positives) и U (unlabeled). Причем брать нужно не все положительные (класс 1) примеры, а только лишь часть

5. применить random negative sampling для построения классификатора в новых условиях

6. сравнить качество с решением из пункта 4 (построить отчет - таблицу метрик)

7. поэкспериментировать с долей P на шаге 6 (как будет меняться качество модели при уменьшении/увеличении размера P)

In [1]:
import pandas as pd
import numpy as np
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import itertools

import matplotlib.pyplot as plt

%matplotlib inline

In [2]:
df = pd.read_csv("churn_data.csv")
df.head(10)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   
5          6    15574012       Chu          645     Spain    Male   44   
6          7    15592531  Bartlett          822    France    Male   50   
7          8    15656148    Obinna          376   Germany  Female   29   
8          9    15792365        He          501    France    Male   44   
9         10    15592389        H?          684    France    Male   27   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   
5       8  113755.78              2          1               0   
6       7       0.00              2          1               1   
7       4  115046.74              4          1               0   
8       4  142051.07              2          0               1   
9       2  134603.88              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0  
5        149756.71       1  
6         10062.80       0  
7        119346.88       1  
8         74940.50       0  
9         71725.73       0

In [3]:
df['Exited'].value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

Давайте построим модель. Сразу же будем работать с использованием sklearn pipeline

In [4]:
#разделим данные на train/test
X_train, X_test, y_train, y_test = train_test_split(df.drop(columns=['Exited']), df['Exited'], random_state=0)

- Категориальные признаки закодируем с помощью OneHotEncoding
- Вещественные оставим пока как есть

In [5]:
#соберем наш простой pipeline, но нам понадобится написать класс для выбора нужного поля
class FeatureSelector(BaseEstimator, TransformerMixin):
    def __init__(self, column):
        self.column = column

    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
        return X[self.column]
    

class NumberSelector(BaseEstimator, TransformerMixin):
    
    def __init__(self, key):
        self.key = key

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        return X[[self.key]]
    
    
class OHEEncoder(BaseEstimator, TransformerMixin):
    def __init__(self, key):
        self.key = key
        self.columns = []

    def fit(self, X, y=None):
        self.columns = [col for col in pd.get_dummies(X, prefix=self.key).columns]
        return self

    def transform(self, X):
        X = pd.get_dummies(X, prefix=self.key)
        test_columns = [col for col in X.columns]
        
        for col_ in self.columns:
            if col_ not in test_columns:
                X[col_] = 0
        return X[self.columns]

In [6]:
df.head(3)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1

Зададим списки признаков

In [7]:
categorical_columns = ['Geography', 'Gender', 'Tenure', 'HasCrCard', 'IsActiveMember']
continuous_columns = ['CreditScore', 'Age', 'Balance', 'NumOfProducts', 'EstimatedSalary']

In [8]:
from sklearn.pipeline import Pipeline
from sklearn.pipeline import FeatureUnion

Теперь нам нужно под каждый признак создать трансформер и объединить их в список.

In [9]:
final_transformers = list()

for cat_col in categorical_columns:
    cat_transformer = Pipeline([
                ('selector', FeatureSelector(column=cat_col)),
                ('ohe', OHEEncoder(key=cat_col))
            ])
    
    final_transformers.append((cat_col, cat_transformer))
    
for cont_col in continuous_columns:
    cont_transformer = Pipeline([
                ('selector', NumberSelector(key=cont_col))
            ])
    
    final_transformers.append((cont_col, cont_transformer))

Объединим все это в единый пайплайн

In [10]:
feats = FeatureUnion(final_transformers)

Теперь у нас есть пайплайн, который готовит признаки для моделирования.

Добавим модель

In [11]:
from sklearn.ensemble import RandomForestClassifier

pipeline = Pipeline([
    ('features', feats),
    ('classifier', RandomForestClassifier(random_state=42)),
])

In [12]:
# обучим наш пайплайн
pipeline.fit(X_train, y_train)

Pipeline(steps=[('features',
                 FeatureUnion(transformer_list=[('Geography',
                                                 Pipeline(steps=[('selector',
                                                                  FeatureSelector(column='Geography')),
                                                                 ('ohe',
                                                                  OHEEncoder(key='Geography'))])),
                                                ('Gender',
                                                 Pipeline(steps=[('selector',
                                                                  FeatureSelector(column='Gender')),
                                                                 ('ohe',
                                                                  OHEEncoder(key='Gender'))])),
                                                ('Tenure',
                                                 Pipeline(steps=[('selector',
           

In [13]:
# наши прогнозы для тестовой выборки
preds = pipeline.predict_proba(X_test)[:, 1]
preds[:10]

array([0.37, 0.26, 0.16, 0.02, 0.02, 0.67, 0.04, 0.12, 0.15, 0.75])

Также нам нужно от вероятностей перейти к меткам классов. Для этого нужно подобрать порог, после которого мы считаем, что объект можно отнести к классу 1 (если вероятность больше порога - размечаем объект как класс 1, если нет - класс 0)

In [14]:
from sklearn.metrics import f1_score, roc_auc_score, precision_score, classification_report, precision_recall_curve, confusion_matrix

In [15]:
metrics_df = pd.DataFrame(columns=['model', 'thresh', 'F-Score', 'Precision', 'Recall', 'ROC AUC'])
metrics_df

Empty DataFrame
Columns: [model, thresh, F-Score, Precision, Recall, ROC AUC]
Index: []

In [16]:
precision, recall, thresholds = precision_recall_curve(y_test, preds)

fscore = (2 * precision * recall) / (precision + recall)
# locate the index of the largest f score
ix = np.argmax(fscore)
print(f'Best Threshold={thresholds[ix]}, F-Score={fscore[ix]:.3f}, Precision={precision[ix]:.3f}, Recall={recall[ix]:.3f}')

Best Threshold=0.38, F-Score=0.641, Precision=0.653, Recall=0.629


In [17]:
roc_auc = roc_auc_score(y_test, preds)
roc_auc

0.8635016710758334

In [18]:
metrics_df = metrics_df.append({
    'model': 'supervised',
    'thresh': thresholds[ix],
    'F-Score': fscore[ix],
    'Precision': precision[ix],
    'Recall': recall[ix],
    'ROC AUC': roc_auc
}, ignore_index=True)

metrics_df

C:\Users\Инна\AppData\Local\Temp\ipykernel_13316\3491017330.py:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  metrics_df = metrics_df.append({


model thresh   F-Score Precision    Recall   ROC AUC
0  supervised   0.38  0.640641  0.653061  0.628684  0.863502

Разделим ваш набор данных на два множества: P (positives) и U (unlabeled). Причем брать нужно не все положительные примеры (класс 1), а только лишь часть

In [19]:
mod_data = X_train.copy()
mod_data['Exited'] = y_train
mod_data.head()

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
2967       2968    15595324    Daniels          579   Germany  Female   39   
700         701    15803457        Hao          750    France  Female   32   
3481       3482    15644686    Kennedy          729     Spain  Female   34   
1621       1622    15777797  Kovalyova          689     Spain    Male   38   
800         801    15747542      Perez          605    France    Male   52   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
2967       5  117833.30              3          0               0   
700        5       0.00              2          1               0   
3481       9   53299.96              2          1               1   
1621       5   75075.14              1          1               1   
800        7       0.00              2          1               1   

      EstimatedSalary  Exited  
2967          5831.00       1  
700          95611.47       0  
3481         42855.97       0  
1621          8651.92       1  
800         173952.50       0

In [20]:
# получить индексы положительных образцов
pos_ind = mod_data[mod_data['Exited'] == 1].sample(frac=1, random_state=42).index

# оставить отмеченными только 25% положительных результатов
perc = 0.25
pos_sample_len = int(np.ceil(perc * len(pos_ind)))

print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 382/1528 as positives and unlabeling the rest


Создаем столбец для новой целевой переменной, где у нас два класса - P (1) и U (-1)

In [21]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample, 'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    7118
 1     382
Name: class_test, dtype: int64


In [22]:
mod_data.head(5)

RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
2967       2968    15595324    Daniels          579   Germany  Female   39   
700         701    15803457        Hao          750    France  Female   32   
3481       3482    15644686    Kennedy          729     Spain  Female   34   
1621       1622    15777797  Kovalyova          689     Spain    Male   38   
800         801    15747542      Perez          605    France    Male   52   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
2967       5  117833.30              3          0               0   
700        5       0.00              2          1               0   
3481       9   53299.96              2          1               1   
1621       5   75075.14              1          1               1   
800        7       0.00              2          1               1   

      EstimatedSalary  Exited  class_test  
2967          5831.00       1          -1  
700          95611.47       0          -1  
3481         42855.97       0          -1  
1621          8651.92       1          -1  
800         173952.50       0          -1

применить random negative sampling для построения классификатора в новых условиях

In [23]:
mod_data = mod_data.sample(frac=1, random_state=42)


data_N = mod_data[mod_data['class_test'] == -1]
data_P = mod_data[mod_data['class_test'] == 1]

neg_sample = data_N[:data_P.shape[0]]
pos_sample = data_P.copy()

print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1, random_state=42)

(382, 15) (382, 15)


In [24]:
sample_train

RowNumber  CustomerId     Surname  CreditScore Geography  Gender  Age  \
2421       2422    15574842     Lorenzo          653   Germany  Female   25   
5828       5829    15684419     Wallace          709     Spain  Female   37   
1117       1118    15592583      Colman          731    France  Female   47   
6303       6304    15591169       Hawes          788   Germany  Female   49   
7645       7646    15698522      Thomas          660   Germany    Male   39   
...         ...         ...         ...          ...       ...     ...  ...   
7042       7043    15617348  Uchechukwu          544    France    Male   44   
9086       9087    15743545      Nworie          647     Spain  Female   29   
6516       6517    15775797    Esposito          607     Spain  Female   32   
7483       7484    15784526        Chen          616    France    Male   44   
6367       6368    15568742      Parkes          536    France  Female   41   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
2421       2  158266.42              3          1               1   
5828       8       0.00              3          1               0   
1117       1  115414.19              3          0               0   
6303       4  137455.99              1          1               0   
7645       9  134599.33              2          1               0   
...      ...        ...            ...        ...             ...   
7042       1       0.00              2          0               0   
9086       2       0.00              2          1               0   
6516       7       0.00              3          0               1   
7483       5  102016.38              1          0               1   
6367       9       0.00              1          1               0   

      EstimatedSalary  Exited  class_test  
2421        199357.24       0          -1  
5828         71738.56       0          -1  
1117        191734.67       1           1  
6303        184178.29       1          -1  
7645        183095.87       0          -1  
...               ...     ...         ...  
7042         69244.24       0          -1  
9086        179032.68       0          -1  
6516         10674.62       0          -1  
7483        178235.37       1           1  
6367        121299.14       0          -1  

[764 rows x 15 columns]

In [25]:
sample_train.loc[sample_train['class_test'] == -1, 'class_test'] = 0

pipeline = Pipeline([
    ('features', feats),
    ('classifier', RandomForestClassifier(random_state=42)),
])


pipeline.fit(sample_train.drop(columns=['class_test', 'Exited']), 
             sample_train['class_test'])

Pipeline(steps=[('features',
                 FeatureUnion(transformer_list=[('Geography',
                                                 Pipeline(steps=[('selector',
                                                                  FeatureSelector(column='Geography')),
                                                                 ('ohe',
                                                                  OHEEncoder(key='Geography'))])),
                                                ('Gender',
                                                 Pipeline(steps=[('selector',
                                                                  FeatureSelector(column='Gender')),
                                                                 ('ohe',
                                                                  OHEEncoder(key='Gender'))])),
                                                ('Tenure',
                                                 Pipeline(steps=[('selector',
           

In [26]:
# прогнозы для тестовой выборки
preds = pipeline.predict_proba(X_test)[:, 1]
preds[:10]

array([0.72, 0.39, 0.58, 0.42, 0.19, 0.8 , 0.15, 0.17, 0.38, 0.6 ])

In [27]:
precision, recall, thresholds = precision_recall_curve(y_test, preds)

fscore = (2 * precision * recall) / (precision + recall)
# locate the index of the largest f score
ix = np.argmax(fscore)
print(f'Best Threshold={thresholds[ix]}, F-Score={fscore[ix]:.3f}, Precision={precision[ix]:.3f}, Recall={recall[ix]:.3f}')

Best Threshold=0.58, F-Score=0.554, Precision=0.489, Recall=0.639


In [28]:
roc_auc = roc_auc_score(y_test, preds)
roc_auc

0.8059154209660564

In [29]:

roc_auc = roc_auc_score(y_test, preds)
roc_auc

0.8059154209660564

In [30]:
#занесем метрики в табличку
metrics_df = metrics_df.append({
    'model': 'pu-learning',
    'thresh': thresholds[ix],
    'F-Score': fscore[ix],
    'Precision': precision[ix],
    'Recall': recall[ix],
    'ROC AUC': roc_auc
}, ignore_index=True)

metrics_df

C:\Users\Инна\AppData\Local\Temp\ipykernel_13316\1445158269.py:2: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  metrics_df = metrics_df.append({


model thresh   F-Score Precision    Recall   ROC AUC
0   supervised   0.38  0.640641  0.653061  0.628684  0.863502
1  pu-learning   0.58  0.554135  0.489458  0.638507  0.805915

сравнить качество с решением из пункта 4 (построить отчет - таблицу метрик)

*Поэкспериментировать с долей P на шаге 5 (как будет меняться качество модели при уменьшении/увеличении размера P)

In [31]:
from tqdm import tqdm

metrics_df = pd.DataFrame(columns=['frac', 'thresh', 'F-Score', 'Precision', 'Recall', 'ROC AUC'])


for frac in tqdm(np.linspace(0.1, 1, 9)):
    mod_data = X_train.copy()
    mod_data['Exited'] = y_train
    mod_data.head()

    # получить индексы положительных образцов
    pos_ind = mod_data[mod_data['Exited'] == 1].sample(frac=1, random_state=42).index

    pos_sample_len = int(np.ceil(frac * len(pos_ind)))
    pos_sample = pos_ind[:pos_sample_len]
    
    mod_data['class_test'] = -1
    mod_data.loc[pos_sample,'class_test'] = 1
    
    mod_data = mod_data.sample(frac=1, random_state=42)


    data_N = mod_data[mod_data['class_test'] == -1]
    data_P = mod_data[mod_data['class_test'] == 1]

    neg_sample = data_N[:data_P.shape[0]]
    pos_sample = data_P.copy()

    sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1, random_state=42)

    sample_train.loc[sample_train['class_test'] == -1, 'class_test'] = 0

    pipeline = Pipeline([
        ('features', feats),
        ('classifier', RandomForestClassifier(random_state=42)),
    ])

    pipeline.fit(sample_train.drop(columns=['class_test','Exited']), 
                 sample_train['class_test'])
    
    # наши прогнозы для тестовой выборки
    preds = pipeline.predict_proba(X_test)[:, 1]
    preds[:10]

    precision, recall, thresholds = precision_recall_curve(y_test, preds)

    fscore = (2 * precision * recall) / (precision + recall)
    # найти индекс наибольшего f-score
    ix = np.argmax(fscore)
    roc_auc = roc_auc_score(y_test, preds)

    metrics_df = metrics_df.append({
        'frac': frac,
        'thresh': thresholds[ix],
        'F-Score': fscore[ix],
        'Precision': precision[ix],
        'Recall': recall[ix],
        'ROC AUC': roc_auc
    }, ignore_index=True)

metrics_df

  0%|                                                                                            | 0/9 [00:00<?, ?it/s]C:\Users\Инна\AppData\Local\Temp\ipykernel_13316\1697198961.py:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  metrics_df = metrics_df.append({
 11%|█████████▎                                                                          | 1/9 [00:00<00:01,  4.55it/s]C:\Users\Инна\AppData\Local\Temp\ipykernel_13316\1697198961.py:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  metrics_df = metrics_df.append({
 22%|██████████████████▋                                                                 | 2/9 [00:00<00:01,  4.58it/s]C:\Users\Инна\AppData\Local\Temp\ipykernel_13316\1697198961.py:52: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. 

frac thresh   F-Score Precision    Recall   ROC AUC
0     0.1    0.6  0.509767  0.484099   0.53831  0.769913
1  0.2125   0.56  0.548037  0.476744  0.644401  0.802129
2   0.325   0.59  0.565062  0.517129   0.62279   0.82408
3  0.4375   0.64  0.577259  0.571154  0.583497  0.832189
4    0.55   0.63  0.586081  0.548885  0.628684  0.839407
5  0.6625   0.64  0.598291  0.579044  0.618861  0.841845
6   0.775   0.63  0.610856  0.574394  0.652259  0.846983
7  0.8875   0.64  0.624309  0.587522  0.666012  0.853069
8     1.0   0.64  0.621072  0.586387  0.660118  0.860818

Качество модели существенно меняется при увеличении доли P (positives) в выборке, причем предел в большинстве случаев одинаков (0.64), precision и recall растут экспененциально, а auc - пропорционально f-score. Оптимальная модель получается при доле positives 0.88